### PR TITLE
Store tracking info in subscriptions

### PR DESCRIPTION
**Pull request by gbirke**
Moved subscription route handling to class.
Add test and code to store subscription source and general tracking
information.

This is for https://github.com/wmde/fundraising/issues/1110

### CODE REVIEW

**Comment by gbirke** [comment]
This will fail until https://github.com/wmde/FundraisingStore/pull/105 is merged and the version requirement in composer.json set to `3.0-dev`

File: contexts/SubscriptionContext/src/UseCases/AddSubscription/AddSubscriptionUseCase.php [hunk 1]
```diff
@@ -82,13 +84,15 @@ private function newMailAddressFromSubscription( Subscription $subscription ): E
...
+		$subscription = new Subscription();
 
-		$request->setAddress( $this->addressFromSubscriptionRequest( $subscriptionRequest ) );
-		$request->setEmail( $subscriptionRequest->getEmail() );
-		$request->setConfirmationCode( random_bytes( 16 ) ); // No need to use uuid library here
+		$subscription->setAddress( $this->addressFromSubscriptionRequest( $subscriptionRequest ) );
+		$subscription->setEmail( $subscriptionRequest->getEmail() );
+		$subscription->setTracking( $subscriptionRequest->getTracking() );
+		$subscription->setSource( $subscriptionRequest->getSource() );
+		$subscription->setConfirmationCode( random_bytes( self::CONFIRMATION_CODE_BYTES ) ); // No need to use uuid library here
```
Comment (JeroenDeDauw) [inline contexts/SubscriptionContext/src/UseCases/AddSubscription/AddSubscriptionUseCase.php]: This const name is a bit misleading; I thought a collection of bytes was being passed in here until I looked at the other side of the diff

File: contexts/SubscriptionContext/src/UseCases/AddSubscription/SubscriptionRequest.php [hunk 2]
```diff
@@ -92,6 +94,22 @@ public function setWikilogin( bool $wikilogin ) {
 		$this->wikilogin = $wikilogin;
 	}
 
+	public function getTracking(): string {
+		return $this->tracking;
```
Comment (JeroenDeDauw) [inline contexts/SubscriptionContext/src/UseCases/AddSubscription/SubscriptionRequest.php]: I'm not very fond of this name. "What is a tracking" is not even correct English. "tracking" is some kind of purpose/context, with the actual noun missing. `trackingString` seems better, and apparently we're already using it at some places https://github.com/wmde/FundraisingFrontend/search?utf8=%E2%9C%93&q=trackingstring. Might be something better still tho

**Comment by JeroenDeDauw** [comment]
Rebased

File: app/RouteHandlers/AddSubscriptionHandler.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		$subscriptionRequest->setLastName( $request->get( 'lastName', '' ) );
+		$subscriptionRequest->setSalutation( $request->get( 'salutation', '' ) );
+		$subscriptionRequest->setTitle( $request->get( 'title', '' ) );
+
+	}
+
+	private function addTrackingDataFromRequest( SubscriptionRequest $subscriptionRequest, Request $request ) {
+		$subscriptionRequest->setSource( $request->get( 'source', '' ) );
+		$subscriptionRequest->setTrackingString( $request->attributes->get( 'trackingCode', '' ) );
+	}
```
Comment (JeroenDeDauw) [inline app/RouteHandlers/AddSubscriptionHandler.php]: Like this its kinda hard to see what you changed. Perhaps its worthwhile to create a commit for the refactor part and one for the new features part. Makes review easier and forces you to not mix the two.

**Comment by JeroenDeDauw** [comment]
CR+2 for what Gabriel did. Now needs someone to review what I added on top.

**Pull request merged**